### ISSUE TITLE
COMBINE STE + DLM continued pre-training into one

### ISSUE THREAD

**Issue by Lev-Stambler** [DeepOpt-com/WrinkleFree-Monorepo#2]
## Problem

Currently, BitNet model training requires **two separate continued pre-training stages**:

1. **Stage 2 STE Training** (`packages/training`) - ~10B tokens
   - Continue pre-training with Straight-Through Estimator (STE)
   - Ternary weight quantization {-1, 0, 1} + 8-bit activations
   - Lambda warmup for gradual quantization
   - Output: Fully quantized 1.58-bit BitNet model

2. **DLM Conversion** (`packages/converter`) - ~1B tokens
   - Fast-dLLM v2 training for parallel inference
   - Response-only SFT loss with block diffusion
   - Adds mask token, learns complementary mask patterns
   - Output: Diffusion LLM with 2.5x faster inference

This is inefficient:
- Two separate training runs = more compute cost
- Two separate data passes = ~11B tokens total
- Two separate optimization phases = potential forgetting between stages

## Proposal

Combine STE quantization training and DLM conversion into a **single unified continued pre-training stage**.

### Continual Quantization-Aware Pre-Training
[arXiv:2502.11895](https://arxiv.org/abs/2502.11895) shows that 16-to-1.58-bit transitions outperform full 1.58-bit training. Optimal transition points are early (~2K steps).

### BitDistill Multi-Stage Pipeline
[arXiv:2510.13998](https://arxiv.org/html/2510.13998) demonstrates that continue-pretraining adapts weight distributions for 1.58-bit optimization before fine-tuning.

### Fast-dLLM v2
Shows that DLM conversion requires only ~1B tokens when starting from a strong base model - suggests objectives can be combined efficiently.

### Phase 1: Multi-Task Loss Function
Create a combined loss that balances:
```python
loss = α * lm_loss + β * diffusion_sft_loss
```

Where:
- `lm_loss`: Standard next-token prediction (Stage 2 STE objective)
- `diffusion_sft_loss`: Response-only SFT with block masks (DLM objective)
- `α, β`: Schedulable weights (can anneal α→0, β→1 over training)

### Phase 2: Unified Data Pipeline
Extend the data loader to support both objectives:
1. Standard pre-training sequences (for `lm_loss`)
2. Conversation/SFT sequences with prompt masking (for `diffusion_sft_loss`)
3. Optional: Interleave or alternate batches

### Phase 3: Combined Lambda Warmup
Single lambda warmup schedule that:
- Gradually enables quantization (λ: 0→1)
- Coordinated with loss weight schedule (α, β)
- Early phase: focus on adaptation (higher α, λ<1)
- Later phase: focus on DLM (higher β, λ→1)

### Phase 4: Unified Trainer
New `Stage2DLMTrainer` that:
- Inherits from both Stage2Trainer and DLM trainer logic
- Handles mask token initialization (smart init to vocab mean)
- Includes ZClip adaptive gradient clipping
- Single checkpoint format with DLM-ready config (bd_size, mask_id)

## File Changes

| File | Change |
|------|--------|
| `packages/training/src/wrinklefree/training/stage2_dlm.py` | **NEW**: Combined trainer |
| `packages/training/configs/training/stage2_dlm.yaml` | **NEW**: Combined config |
| `packages/training/src/wrinklefree/losses/combined.py` | **NEW**: Multi-task loss |
| `packages/cheapertraining/src/cheapertraining/data/sft.py` | Add conversation dataset support |

## Success Criteria

1. Single training run produces DLM-ready 1.58-bit model
2. Total tokens ≤ current combined (target: ~5-8B vs 11B)
3. No regression in downstream task performance
4. Inference speedup from DLM preserved (2.5x)

## Open Questions

- [ ] Optimal α/β schedule - constant vs annealing?
- [ ] Should LM loss use bidirectional attention like DLM or stay causal?
- [ ] Data mixing ratio for pretrain vs SFT sequences?
- [ ] Should we share the mask token initialization with vocab mean?

## References

- [Continual QAT Pre-Training](https://arxiv.org/abs/2502.11895)
- [BitDistill](https://arxiv.org/html/2510.13998)
- [Fast-dLLM v2](https://arxiv.org/abs/2509.26328)
- [STE Theory](https://arxiv.org/abs/1903.05662)
- [HuggingFace 1.58-bit Fine-tuning](https://huggingface.co/blog/1_58_llm_extreme_quantization)
